cmd/acra-server/common/http_api.go

```diff
@@ -0,0 +1,310 @@
+// Copyright 2022, Cossack Labs Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"io/ioutil"
+	stdlog "log"
+	"net"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"github.com/cossacklabs/acra/logging"
+	"github.com/cossacklabs/acra/network"
+	"github.com/cossacklabs/acra/zone"
+	"github.com/cossacklabs/themis/gothemis/keys"
+	"github.com/gin-gonic/gin"
+	"github.com/gin-gonic/gin/render"
+	log "github.com/sirupsen/logrus"
+	"go.opencensus.io/trace"
+)
+
+const (
+	errorRequestMessage = "incorrect request"
+	// Key name for the logger in the gin.Context
+	loggerKey = "logger"
+	// Key name for the clientID saved in the gin.Context
+	clientIDKey = "clientID"
+)
+
+// AcraAPIServer handles all HTTP api logic
+type AcraAPIServer struct {
+	ctx        context.Context
+	api        APICore
+	engine     *gin.Engine
+	httpServer *http.Server
+}
+
+// APICore contains the API logic of the HTTP API server
+// Is used to decouple the API logic from actual HTTP setting routine
+// In the future could be used to abstract HTTP setting up from API configuring
+type APICore struct {
+	server *SServer
```

Comment:
for what apiCore needs server? if it is depends on server, then maybe better to implement api through server? otherwise lets do it server agnostic. as I see, it is depends on keystore or config. so lets initialize it with config or keystore instance instead of server. then we can re-use it for other kind of server (gRPC for example) as we do it in AT module

Reply:
I thought that it can be useful if we will add more endpoints that will use more features of the SSserver. But yeah, decoupling is better and easier to test.